### PR TITLE
[pytest/hash] Add hash keys: 'src-mac', 'dst-mac', 'ip-proto', 'vlan-id'.

### PR DESCRIPTION
**Pull request by William-zx**
### Description of PR

Summary:
Made the following change:

Add some hash keys: 'src-mac', 'dst-mac', 'ip-proto', 'vlan-id'.

Sub PR of [Enhance hash test #1494](https://github.com/Azure/sonic-mgmt/pull/1494)

This PR base on the following PR:

[Rewrite hash test #1506](https://github.com/Azure/sonic-mgmt/pull/1494)

[Complete hash key ingress-port test #1509](https://github.com/Azure/sonic-mgmt/pull/1509)

[Generate available send packet ports instead of hard coding src_ports in hash_test.py. #1511](https://github.com/Azure/sonic-mgmt/pull/1511)

### Type of change

- [] Bug fix
- [] Testbed and Framework(new/improvement)
- [x] Test case(new/improvement)

### CODE REVIEW

**Comment by lguohan** [comment]
can you rebase this and I'll do code review on this pr?

**Comment by William-zx** [comment]
> can you rebase this and I'll do code review on this pr?

Conflict resolved, please review.

File: ansible/roles/test/files/ptftests/hash_test.py [hunk 1]
```diff
@@ -48,14 +48,17 @@ def setUp(self):
...
         self.fib = fib.Fib(self.test_params['fib_info'])
+        self.testbed_type = self.test_params['testbed_type']
         self.router_mac = self.test_params['router_mac']
         self.in_ports = self.test_params['in_ports']
 
         self.src_ip_range = [unicode(x) for x in self.test_params['src_ip_range'].split(',')]
         self.dst_ip_range = [unicode(x) for x in self.test_params['dst_ip_range'].split(',')]
         self.src_ip_interval = lpm.LpmDict.IpInterval(ip_address(self.src_ip_range[0]), ip_address(self.src_ip_range[1]))
         self.dst_ip_interval = lpm.LpmDict.IpInterval(ip_address(self.dst_ip_range[0]), ip_address(self.dst_ip_range[1]))
+        self.vlans = self.test_params.get('vlans', [])
```
Comment (lguohan) [inline ansible/roles/test/files/ptftests/hash_test.py:59]: can you rename to vlan_ids for more clarity.

Comment (William-zx) [reply ansible/roles/test/files/ptftests/hash_test.py]: done.

File: tests/fib/test_fib.py [hunk 2]
```diff
@@ -1,19 +1,19 @@
...
 import json
 import logging
-import requests
 from ptf_runner import ptf_runner
 from datetime import datetime 
 
 logger = logging.getLogger(__name__)
 
-HASH_KEYS = ['src-ip', 'dst-ip', 'src-port', 'dst-port', 'ingress-port']
+HASH_KEYS = ['src-ip', 'dst-ip', 'src-port', 'dst-port', 'ingress-port', 'src-mac', 'dst-mac', 'ip-proto', 'vlan-id']
```
Comment (lguohan) [inline tests/fib/test_fib.py:10]: usually src-mac, dst-mac, vlan-id are optional hash keys. Not all the platform supports these optional hash keys. I think we need to provide options for the user to choose to test which hash key.

Comment (lguohan) [reply tests/fib/test_fib.py:10]: a simple fix is to not enable these three by default. ideally, I think we should break each hash field into one test. and then we can use @pytest.mark.skip to skip running the tests on certain platform (long term)

File: tests/fib/test_fib.py [hunk 3]
```diff
@@ -158,6 +161,26 @@ def setup_hash(self, testbed, duthost, ptfhost):
         in_ports_name = get_router_interface_ports(config_facts, testbed)
         g_vars['in_ports'] = [config_facts.get('port_index_map', {})[p] for p in in_ports_name]
 
+        # add some vlan for hash_key vlan-id test
```
Comment (lguohan) [inline tests/fib/test_fib.py:164]: this setup needs to move to a separate fixture, for example, setup_vlan. Then, only test_hash_key_vlan_id load this fixture.

File: tests/fib/test_fib.py (hunk 2 above)
Comment (William-zx) [reply tests/fib/test_fib.py]: > a simple fix is to not enable these three by default. ideally, I think we should break each hash field into one test. and then we can use @pytest.mark.skip to skip running the tests on certain platform (long term)

Not enable src-mac, dst-mac and vlan-id hash keys by default.

File: tests/fib/test_fib.py (hunk 3 above)
Comment (William-zx) [reply tests/fib/test_fib.py:177]: > this setup needs to move to a separate fixture, for example, setup_vlan. Then, only test_hash_key_vlan_id load this fixture.

Add a determination of whether vlan-id is in hash_keys.

Move setup_vlan to a separate fixture for only test_hash_key_vlan_id load this fixture. That means break each hash field into one test, should we change it in this PR?

Comment (lguohan) [reply tests/fib/test_fib.py:177]: it is better to separate each hash field into one test. you can do it in future prs.

**Pull request merged**